networkx/algorithms/centrality/group.py

```diff
@@ -84,49 +84,196 @@ def group_betweenness_centrality(G, C, normalized=True, weight=None):
+                            * sigma[group_node1][group_node2]
+                            / sigma[node][group_node2]
+                        )
+
+    # the algorithm for each group
+    for group in C:
+        group = set(group)  # set of nodes in group
+        # initialize the matrices of the sigma and the PB
+        GBC_group = 0
+        sigma_m = deepcopy(sigma)
+        PB_m = deepcopy(PB)
+        sigma_m_v = deepcopy(sigma_m)
+        PB_m_v = deepcopy(PB_m)
+        for v in group:
+            GBC_group += PB_m[v][v]
+            for x in group:
+                for y in group:
+                    dxvy = 0
+                    dxyv = 0
+                    dvxy = 0
+                    if not (
+                        sigma_m[x][y] == 0 or sigma_m[x][v] == 0 or sigma_m[y][v] == 0
+                    ):
+                        if D[x][v] == D[x][y] + D[y][v]:
+                            dxyv = 1.0 * sigma_m[x][y] * sigma_m[y][v] / sigma_m[x][v]
+                        if D[x][y] == D[x][v] + D[v][y]:
+                            dxvy = 1.0 * sigma_m[x][v] * sigma_m[v][y] / sigma_m[x][y]
+                        if D[v][y] == D[v][x] + D[x][y]:
+                            dvxy = 1.0 * sigma_m[v][x] * sigma[x][y] / sigma[v][y]
+                    sigma_m_v[x][y] = sigma_m[x][y] * (1 - dxvy)
+                    PB_m_v[x][y] = PB_m[x][y] - PB_m[x][y] * dxvy
+                    if y != v:
+                        PB_m_v[x][y] -= PB_m[x][v] * dxyv
+                    if x != v:
+                        PB_m_v[x][y] -= PB_m[v][y] * dvxy
+            tmp = PB_m
+            PB_m = PB_m_v
+            PB_m_v = tmp
+            tmp = sigma_m
+            sigma_m = sigma_m_v
+            sigma_m_v = tmp
+
+        # endpoints
+        v, c = len(G), len(group)
+        if not endpoints:
+            scale = 0
+            # if the graph is connected than deduce the endpoints of all the nodes in the graph. else find which nodes
+            # are connected to the group's nodes and deduce them
+            if nx.is_directed(G):
+                if nx.is_strongly_connected(G):
+                    scale = c * (2 * v - c - 1)
+            elif nx.is_connected(G):
+                scale = c * (2 * v - c - 1)
+            if scale == 0:
+                for group_node1 in group:
+                    for node in G:
+                        if D[group_node1][node] != float("inf") and node != group_node1:
+                            if node in group:
+                                scale += 1
+                            else:
+                                scale += 2
+            GBC_group -= scale
+
+        # normalized
+        if normalized:
+            scale = 1 / ((v - c) * (v - c - 1))
+            if not G.is_directed():
+                scale *= 2
+            GBC_group *= scale
+
+        GBC.append(GBC_group)
+    return GBC
+
+
+def _single_source_shortest_path_basic(G, s):
```

Comment:
I think you can use the helper functions from betweenness.py as suggested by @rossbar becuse that function returns `S, P, sigma` and then you can build delta using the code after your comment: `accumulate`.

Reply:
I also need to get the dictionary  `D` (the distance from s to all other nodes) from `_single_source_shortest_path_basic` function. Should I change the function in betweenness.py, keep the helper functions in group.py, or calculate D from `S, P, sigma` (which would increase the running time of the algorithm)?

Reply:
I think `dist` is already calculated in the helper functions in `betweenness.py` so we should change those two helper functions to return `S, pred, sigma, dist` ... which will require changes in every function that uses these helper functions -- which I believe is once(for each function) in `betweenness.py`, once in `percolation.py` and once in `betweenness_subset.py`.

Then in `group.py` you can call it and compute `delta`.  Does this make sense?

Reply:
I change the helper functions in betweenness.py so they return delta and D. I also change each function which uses those helper functions.
Now the group betweenness centrality function calls the helper functions from betweenness.